core/src/main/java/jenkins/telemetry/impl/AutoRefresh.java

```diff
@@ -66,16 +66,19 @@ public LocalDate getStart() {
     @Nonnull
     @Override
     public LocalDate getEnd() {
-        return LocalDate.of(2019, 4, 10);
+        return LocalDate.of(2019, 5, 31);
     }
 
     @Override
     public JSONObject createContent() {
         if (sessionsBySetting.size() == 0) {
             return null;
         }
-        Map<Boolean, Set<String>> currentSessions = new TreeMap<>(sessionsBySetting);
-        sessionsBySetting.clear();
+        Map<Boolean, Set<String>> currentSessions;
+        synchronized (sessionsBySetting) {
+            currentSessions = new TreeMap<>(sessionsBySetting);
+            sessionsBySetting.clear();
+        }
```

Comment:
Does not look to be enough, because `currentSessions.entrySet()` iterator below may still collide with `recordRequest()`. I recommend synchronizing the entire block

Reply:
It's a different collection now. All that would happen is that requests arriving after we've started preparing the `JSONObject` will be counted towards the next iteration/submission, a day later. Or am I missing something?

Would rather not block request processing while we're building a potentially large `JSONObject` over several seconds.

Reply:
@daniel-beck Since this is a eventual-consistency call, you might want to separate collections: recording in a sing collection and then copying/switching values in the beginning of the reporting call. Then this long procedure will be operating with a local snapshot which is not longer locked

Reply:
I do not want to waste the next hour of my life on a blackboard tracing out the possible race conditions here. Can you slap a `synchronized` block around all accesses in `createContent` and `recordRequest` and use simple, obviously correct code please?

Reply:
> you might want to separate collections: recording in a sing collection and then copying/switching values in the beginning of the reporting call. Then this long procedure will be operating with a local snapshot which is not longer locked

@oleg-nenashev But that's exactly what this code is doing? `new TreeMap<>(sessionsBySetting)` is the same as `new TreeMap<>().addAll(sessionsBySetting)`, which is why I can clear the original list immediately afterwards. Below I only operate on the local copy of the map.

Reply:
🤦‍♂ Forgot about the nested sets.